src/cpu/Dasm.cc

```diff
@@ -23,96 +23,122 @@ void appendAddrAsHex(std::string& output, uint16_t addr)
-              function_ref<void(std::string&, uint16_t)> appendAddr)
+unsigned instructionLength(std::span<const uint8_t> bin)
+{
+	if (bin.empty()) return 0;
+	auto t = instr_len_tab[bin[0]];
+	if (t < 4) return t;
+
+	if (bin.size() < 2) return 0;
+	return instr_len_tab[64 * t + bin[1]];
+}
+
+void dasm(std::span<const uint8_t> bin, uint16_t pc, std::string& dest,
+          function_ref<void(std::string&, uint16_t)> appendAddr)
 {
 	const char* r = nullptr;
 
-	buf[0] = interface.peekMem(pc, time);
 	auto [s, i] = [&]() -> std::pair<const char*, unsigned> {
-		switch (buf[0]) {
+		switch (bin[0]) {
 			case 0xCB:
-				buf[1] = interface.peekMem(pc + 1, time);
-				return {mnemonic_cb[buf[1]], 2};
+				return {mnemonic_cb[bin[1]], 2};
 			case 0xED:
-				buf[1] = interface.peekMem(pc + 1, time);
-				return {mnemonic_ed[buf[1]], 2};
+				return {mnemonic_ed[bin[1]], 2};
 			case 0xDD:
 			case 0xFD:
-				r = (buf[0] == 0xDD) ? "ix" : "iy";
-				buf[1] = interface.peekMem(pc + 1, time);
-				if (buf[1] != 0xcb) {
-					return {mnemonic_xx[buf[1]], 2};
+				r = (bin[0] == 0xDD) ? "ix" : "iy";
+				if (bin[1] != 0xcb) {
+					return {mnemonic_xx[bin[1]], 2};
 				} else {
-					buf[2] = interface.peekMem(pc + 2, time);
-					buf[3] = interface.peekMem(pc + 3, time);
-					return {mnemonic_xx_cb[buf[3]], 4};
+					return {mnemonic_xx_cb[bin[3]], 4};
 				}
 			default:
-				return {mnemonic_main[buf[0]], 1};
+				return {mnemonic_main[bin[0]], 1};
 		}
 	}();
 
-	for (int j = 0; s[j]; ++j) {
-		switch (s[j]) {
-		case 'B':
-			buf[i] = interface.peekMem(narrow_cast<uint16_t>(pc + i), time);
-			strAppend(dest, '#', hex_string<2>(
-				static_cast<uint16_t>(buf[i])));
-			i += 1;
-			break;
-		case 'R':
-			buf[i] = interface.peekMem(narrow_cast<uint16_t>(pc + i), time);
-			appendAddr(dest, uint16_t(pc + 2 + static_cast<int8_t>(buf[i])));
-			i += 1;
-			break;
-		case 'A':
-		case 'W':
-			buf[i + 0] = interface.peekMem(narrow_cast<uint16_t>(pc + i + 0), time);
-			buf[i + 1] = interface.peekMem(narrow_cast<uint16_t>(pc + i + 1), time);
-			appendAddr(dest, buf[i] + buf[i + 1] * 256);
-			i += 2;
-			break;
-		case 'X':
-			buf[i] = interface.peekMem(narrow_cast<uint16_t>(pc + i), time);
-			strAppend(dest, '(', r, sign(buf[i]), '#',
-			     hex_string<2>(abs(buf[i])), ')');
-			i += 1;
-			break;
-		case 'Y':
-			strAppend(dest, r, sign(buf[2]), '#', hex_string<2>(abs(buf[2])));
-			break;
-		case 'I':
-			dest += r;
-			break;
-		case '!':
-			dest = strCat("db     #ED,#", hex_string<2>(buf[1]),
-			              "     ");
-			return 2;
-		case '@':
-			dest = strCat("db     #", hex_string<2>(buf[0]),
-			              "         ");
-			return 1;
-		case '#':
-			dest = strCat("db     #", hex_string<2>(buf[0]),
-			              ",#CB,#", hex_string<2>(buf[2]),
-			              ' ');
-			return 2;
-		case ' ': {
-			dest.resize(7, ' ');
-			break;
-		}
-		default:
-			dest += s[j];
-			break;
+	i = [&]() -> unsigned {
```

Comment:
Maybe I'm overlooking something, but why did you turn this into a lambda?
This new lambda captures `i`, modifies it (it modifies the same `i` variable outside the lambda), returns it, and then assigns it to the same `i` variable (so this assignment does nothing). Unless I missed something I propose to revert this change.

Maybe you did this for symmetry with the code above that initializes `s` and `i`? But that lambda has a different purpose. Without that lambda the code should have been:
```
const char* s; // uninitialized
unsigned i; // uninitialized
switch (...) {
    case ...:
        s = ...;
        i = ...;
}
```
So this is a way to avoid having uninitialized variables (and have the compiler help enforce this). But this is a tradeoff, the lambda does make the code more complex.

Reply:
Aha, I indeed missed that.

I don't like that, I mean my own code. Maybe we can make the switch more regular? So not have a mix between `break` and `return` in the different cases, but only use `break`. The `@`, `#` and `!` are always the last character in the format string, so it should be possible.

Of course we have to think about what to do exactly with the `return` statement. One possibility is to replace `return <n>` with `i = <n>; break;`
* For `!` (only used for `ED`) we can also leave out `i = 2;`, it already has this value, at the time `!` is processed.
* For `@` (used for `DD/FD` instructions unrelated to the `IX/IY` register) the `i = 1;` assignment is needed.
* For `#` the current implementation is weird. We print `db #DD,#CB,#nn` (so 3 bytes), but then only report an instruction length of 2 (`i = 2;`). I think that's a bug, we should either only print 2 bytes (my preference) or skip over 3 instead of 2 bytes.

This is a weird structure and a bug in the existing code. Feel free to tackle it in this PR, but it's also fine if you leave it for later.

Background info: All 3 of `!`, `@` and `#` are only used to disassemble _invalid_ instructions. Our implementation in `CPUCore` emulates these in the same way as how a real Z80 would handle them. But that's _not_ the same as how this `dasm()` function prints them. (Practical problem: the real Z80 has no upper limit on the instruction length (at least not for invalid instructions, e.g. DD DD DD DD DD ... DD 00). Taking that into account in the `dasm()` function introduces extra complexity for no practical gain.)

Reply:
Now that I think about it some more. In case of invalid instructions that `assert(i == bin.size())` will not always be true.

For example `DD CB nn 00` will (currently) be disassembled as `db #DD,#CB,#nn` and will skip over 2 bytes (inconsistent). BUT to be able to determine that this is an invalid instruction we DO need to look at 4 bytes in `bin` (we need to see this `00` value).

And something similar happens for `@`, we only report a length of 1 byte (and print e.g. `db #DD`), but we needed to look at 2 bytes in `bin` to determine that.

So maybe we should skip the assert in case of invalid instructions (and then the `return` in the switch was not such a bad idea ... but an extra comment explaining this would be nice). We should also fix the content of the `instr_len_tab[]` array so that it does fetch enough bytes so that `dasm()` won't crash on invalid instructions (= won't read past the end of `bin`).